### ISSUE TITLE
global is undefined in a strict environment

### ISSUE THREAD

**Issue by EmilTholin** [medikoo/es5-ext#89]
The following code will be generated when using the `es6-set` polyfill in a Rollup IIFE build:

```javascript
(function() {
  "use strict";

  var isImplemented = function() {
    // ...
  };
  /* eslint strict: "off" */
  var global = (function() {
    return this;
  })();

  if (!isImplemented()) {
    Object.defineProperty(global, "Set", {
      value: polyfill$1,
      configurable: true,
      enumerable: false,
      writable: true
    });
  }
})();

```

`global` will be `undefined` because of the `"use strict";`, and the error `Object.defineProperty: argument is not an Object` will be thrown.

Is the recommended solution to remove the `"use strict";`, or could perhaps the global implementation be changed to something along the lines of the implementation in `core-js`?

```javascript
// https://github.com/zloirock/core-js/issues/86#issuecomment-115759028
var global =
  typeof window == "object" && window && window.Math == Math
    ? window
    : typeof self == "object" && self && self.Math == Math
    ? self
    : Function("return this")();
```

**Comment by mathiasbynens** [comment]
I'll just leave this here: https://mathiasbynens.be/notes/globalthis

**Comment by EmilTholin** [comment]
That's ~~a good~~ an *interesting* example, indeed. Fantastic write-up.

**Comment by medikoo** [comment]
> Is the recommended solution to remove the "use strict";

This is CJS module, and for this to work, bundler (or its configuration) should respect a CJS environment characteristics which is non-strict by default.

> or could perhaps the global implementation be changed to something along the lines of the implementation in core-js

It was a bit like that, but then it rose CSP concern for some users, so it was changed.

Technically current implementation is simple to the bone and perfectly valid as far as CJS module is concerned

**Comment by EmilTholin** [comment]
Thank you for your answer and insight.

What do you think about the possibility to implement something along the lines of what is outlined in the [blog post linked above](https://mathiasbynens.be/notes/globalthis), which doesn't have the mentioned CSP issue?

```javascript
var global = (function() {
  if (typeof globalThis === "object") return globalThis;
  try {
    Object.defineProperty(Object.prototype, "__magic__", {
      get: function() {
        return this;
      },
      configurable: true
    });

    var result = __magic__;
    delete Object.prototype.__magic__;
    
    return result;
  } catch (error) {
    // In IE8, Object.defineProperty only works on DOM objects.
    // If we hit this code path, assume `window` exists.
    return window;
  }
})();
```

**Comment by medikoo** [comment]
It'll also work, but question why do bother with such verbose hacks, if simple `module.exports = this` works on its own?

**Comment by EmilTholin** [comment]
It would increase the number of environments this great library can be used in, and subsequently increase the number of environments where `es6-set` et al. can be used.

If you don't want it, I totally understand. I will just use other polyfills instead.

**Comment by zloirock** [comment]
@medikoo CSP is not an issue since in engines where it's possible we check `window` and `self` before usage of `Function` constructor.

**Comment by medikoo** [comment]
> It would increase the number of environments this great library can be used in

Technically It's a (N)CJS module, which in non Node.js envs requires CJS compliant bundling. Purely for simplicity I didn't plan to go beyond that.

Why actually you have problems bundling this module with Rollup? Is (N)CJS support in Rollup not fully compliant?

**Comment by EmilTholin** [comment]
I can set the Rollup output `strict` settings to `false` and this module works great like you said, but it's not immediately obvious that that's the case and I had to spend some time to figure out what was wrong when using `es6-set`, and I'm sure I'm not the last one that will do this mistake.

Again, if you want to keep this a CJS compliant module and not worry about other environments, that's perfectly fine.

**Comment by medikoo** [comment]
> I can set the Rollup output `strict` settings to `false`

If it's a default for Rollup, then I think it should be reported over there. I doesn't sound reasonable to break CJS env for CJS modules by default.
It should rather be an opt-in that comes with a proper notice.

**Comment by medikoo** [comment]
As I looked into Rollup, this choice comes from fact that result bundle is prepared as an ESM module, which is `strict` by design.

This brings some limitations when CJS support is concerned, still has valid grounds, I'll reopen this for reconsideration

**Comment by medikoo** [comment]
Addressed with https://github.com/medikoo/es5-ext/commit/c6a19d703886e300270329552c361a4c0aa33927, published with v0.10.50

Thanks @mathiasbynens for hint on neat getter hack!
